relay-server/src/processing/spans/mod.rs

```diff
@@ -365,33 +469,195 @@ pub struct Indexed;
+            }
+        }
+
+        if attachment > 0 {
+            let limits = rate_limiter
+                .try_consume(scoping.item(DataCategory::Attachment), attachment)
+                .await;
+
+            if !limits.is_empty() {
+                self.modify(|this, record_keeper| {
+                    // Reject both stand_alone and associated attachments.
+                    let mut all_attachments = std::mem::take(&mut this.stand_alone_attachments);
+                    all_attachments.extend(
+                        this.spans
+                            .iter_mut()
+                            .filter_map(|span| span.attachment.take()),
+                    );
+
+                    record_keeper.reject_err(Error::from(limits), all_attachments);
+                });
+            }
+        }
+
+        if attachment_item > 0 {
+            let limits = rate_limiter
+                .try_consume(scoping.item(DataCategory::AttachmentItem), attachment_item)
+                .await;
+
+            if !limits.is_empty() {
+                self.modify(|this, record_keeper| {
+                    // Reject both stand_alone and associated attachments.
+                    let mut all_attachments = std::mem::take(&mut this.stand_alone_attachments);
+                    all_attachments.extend(
+                        this.spans
+                            .iter_mut()
+                            .filter_map(|span| span.attachment.take()),
+                    );
+
+                    record_keeper.reject_err(Error::from(limits), all_attachments);
+                });
+            }
+        }
+
+        Ok(())
+    }
 }
 
 /// A Span which only represents the indexed category.
 #[cfg(feature = "processing")]
 #[derive(Debug)]
-struct IndexedSpan(crate::envelope::WithHeader<SpanV2>);
+struct IndexedSpan(SpanWrapper);
 
 #[cfg(feature = "processing")]
 impl Counted for IndexedSpan {
     fn quantities(&self) -> Quantities {
-        smallvec::smallvec![(DataCategory::SpanIndexed, 1)]
+        let mut quantities = smallvec::smallvec![(DataCategory::SpanIndexed, 1)];
+        if let Some(attachment) = &self.0.attachment {
+            quantities.extend(attachment.quantities());
+        }
+        quantities
+    }
+}
+
+/// A validated and parsed span attachment.
+#[derive(Debug)]
+pub struct ValidatedSpanAttachment {
+    /// The parsed metadata from the attachment.
+    pub meta: Annotated<AttachmentV2Meta>,
+
+    /// The raw attachment body.
+    pub body: Bytes,
+}
+
+impl Counted for ValidatedSpanAttachment {
+    fn quantities(&self) -> Quantities {
+        smallvec::smallvec![
+            (DataCategory::Attachment, self.body.len()),
+            (DataCategory::AttachmentItem, 1)
+        ]
+    }
+}
+
+/// Wrapper around a SpanV2 and an optional associated attachment.
+///
+/// Allows for dropping the attachment together with the Span.
+#[derive(Debug)]
+struct SpanWrapper {
```

Comment:
Some more bikeshedding on names, tell me what you think.

```suggestion
struct ExpandedSpan {
```